Feature/stable repo links

chrpr opened the pull request:
Hey @barnabyalter , @NYULibraries/finding-aids 

I've deleted, re-based, and re-pushed my branch with a re-written version of this feature. This PR replaces #24 .

I think I addressed all of Barnaby's comments in that thread, and all my tests are passing.

I tried pulling out the coffeescript lines per Barnaby's suggestion, but the routes definitely don't affect the dropdown menu, just the actual search values.

I think I implemented to letter the feature request, but I have a question about the functionality for the product owners regarding a consequence of that implementation. I suspect that this will result in a new feature request. I'll put that info in Pivotal rather than here, though.

Let me know what you think, and whether you have any additional comments.

Thanks!

chrpr commented:
FYI -- I have no idea why that travis build failed. It's fine under the other 2 environments, and on my local setup. I don't believe I touched that feature at all.

barnabyalter commented on app/assets/javascripts/findingaids.js.coffee:
```diff
@@ -1,4 +1,5 @@
   # Hide search fields if url specifies repository
   if $("#repository_header").html() != ""
-    $("#search_field").hide()
+    $("#search_field option[value='" + $('#repository_header').text().replace(': ', '') + "']").attr("selected","selected");
```
@chrpr You actually do not need this, setting `:search_field` in the route will preselect this. You'll see that it works for all the libraries except the ones with ampersands because there is a URL encoding clash. So there are a few options here: to keep this functionality you could add something like the following the repositories YAML:

```
url_safe_display: 'The Fales Library %26 Special Collections'
```

And then make that the value of `:search_field` both in the route and in the catalog_controller where you're setting `add_search_field`.

But looking through this functionality there is an inherent problem that I believe has been causing what seems like a bug (but it's actually a feature :smile: ): When you are at any stable repos url, e.g. `/rism`, the `:search_field` will always be set to 'Research Institute for the Study of Man'. This means that if you select Fales from the dropdown then it will still search RISM.

To fix this we will need to use the coffee but we need to allow it to be changed based on selection. The way you have it now forces it to always be selected based on the header text. So something like:

```
if $("#repository_header").html() != "" && $( "#search_field option:selected" ).text() == "All Libraries"
    $("#search_field option[value='" + $('#repository_header').text().replace(': ', '') + "']").attr("selected","selected");
```

barnabyalter commented on app/helpers/results_helper.rb:
```diff
@@ -24,8 +24,13 @@ def render_series_facet_link(doc)
+    repository_label repositories.find{|key,hash| hash["admin_code"] == doc}[1]["url"]
+  end
+  
+  def render_repository_link(doc)
```
@chrpr Please provide specs for these functions as they are the reason code coverage is going down: https://coveralls.io/builds/2122685/source?filename=app%2Fhelpers%2Fresults_helper.rb

barnabyalter commented on config/routes.rb:
```diff
@@ -6,7 +6,8 @@
   YAML.load_file( File.join(Rails.root, "config", "repositories.yml") )["Catalog"]["repositories"].each do |coll|
-     get "#{coll[0]}" => "catalog#index", :search_field => "#{coll[1]['display']}", :repository => "#{coll[1]['display']}"
+     get "#{coll[1]['url']}" => "catalog#index", :search_field => "#{coll[1]['display']}", :repository => "#{coll[1]['display']}"
+
```
@chrpr As mentioned above you'll need to remove the search_field and you can use the Repositories class to DRY it up (https://github.com/NYULibraries/findingaids/blob/development/lib/findingaids/repositories.rb):

```ruby
Findingaids::Repositories.repositories.each do |coll|
    get "#{coll.last['url']}" => "catalog#index", :repository => "#{coll.last['display']}"
end
```

If you feel like it you can also build out the Repositories class so that each repository in `.repositories` is a `Repository` and so you can query `Repositories.repositories.each {|repository| p repository.display}` etc.

chrpr commented:
Thanks @barnabyalter That's too funny, re: URL encoding. I didn't track down the true source of the problem because we only have data in ampersand-containing collection names. Good catch.

I've fixed this issue, rewritten the code a bit, and added the helper function specs you asked for. Can you have a look and see what you think?

Also, re the bug/feature, you're absolutely right. That's actually the issue that I was going to raise with the product owners. My initial idea had been to use the coffee to set the form action url to the correct route based on select, but I think I like where you're going with it better. 

I'm now working on that coffeescript, and am a bit confused by your example. It looks like this says to set the dropdown selected value to whatever is in the repository header, but only if the current selection is "All Libraries"? I'm just not sure what that would do... I also don't really follow what you mean by, "As mentioned above you'll need to remove the search_field and you can use the Repositories class to DRY it up" Where do you mention removing the search_field above? Do you mean removing it from the coffeescript?

Sorry to be a bit dense, and thanks again for your feedback.

barnabyalter commented:
@chrpr Remove the `:search_field` param from the route. Then only select the dropdown from the `:repository` param if no `:search_field` has been selected, i.e. 'All Libraries'. If you then select another dropdown option and submit the form the `:search_field` param will be set to that new value and so the coffeescript will not force the select to be the repos. 

Unless of course desired functionality is that when you select an option from the dropdown and hit search that the default view becomes that repos, which makes sense but is not in the acceptance criteria.

chrpr commented:
Closing this PR & replacing with a fresh one that reflects discussion at last sprint planning.
